### ISSUE TITLE
Error raised inside the event loop: no implicit conversion of Puma::MiniSSL::Socket into Integer

### ISSUE THREAD

**Issue by rda1902** [puma/puma#865]
I use rails actioncable, which include puma. Trying to make puma to work with ssl. 
I run command
"bundle exec puma -b 'ssl://0.0.0.0:28080?key=/var/ssl/exams-st.key&cert=/var/ssl/exams-st.crt' /home/deployer/exams/current/cable/config.ru"

and get error 
```
Puma starting in single mode...
Version 2.15.3 (ruby 2.2.1-p85), codename: Autumn Arbor Airbrush
Min threads: 0, max threads: 16
Environment: development
Listening on ssl://0.0.0.0:28080?key=/var/ssl/exams-st.key&cert=/var/ssl/exams-st.crt
Use Ctrl-C to stop
Error raised inside the event loop: no implicit conversion of Puma::MiniSSL::Socket into Integer
/home/deployer/exams/shared/bundle/ruby/2.2.0/gems/eventmachine-1.0.9.1/lib/eventmachine.rb:767:in `attach_fd'
/home/deployer/exams/shared/bundle/ruby/2.2.0/gems/eventmachine-1.0.9.1/lib/eventmachine.rb:767:in `attach_io'
/home/deployer/exams/shared/bundle/ruby/2.2.0/gems/eventmachine-1.0.9.1/lib/eventmachine.rb:745:in `attach'
/home/deployer/exams/shared/bundle/ruby/2.2.0/gems/faye-websocket-0.10.2/lib/faye/rack_stream.rb:37:in `block in hijack_rack_socket'
/home/deployer/exams/shared/bundle/ruby/2.2.0/gems/eventmachine-1.0.9.1/lib/eventmachine.rb:258:in `call'
/home/deployer/exams/shared/bundle/ruby/2.2.0/gems/eventmachine-1.0.9.1/lib/eventmachine.rb:258:in `block in schedule'
/home/deployer/exams/shared/bundle/ruby/2.2.0/gems/eventmachine-1.0.9.1/lib/eventmachine.rb:981:in `call'
/home/deployer/exams/shared/bundle/ruby/2.2.0/gems/eventmachine-1.0.9.1/lib/eventmachine.rb:981:in `block in run_deferred_callbacks'
/home/deployer/exams/shared/bundle/ruby/2.2.0/gems/eventmachine-1.0.9.1/lib/eventmachine.rb:978:in `times'
/home/deployer/exams/shared/bundle/ruby/2.2.0/gems/eventmachine-1.0.9.1/lib/eventmachine.rb:978:in `run_deferred_callbacks'
/home/deployer/exams/shared/bundle/ruby/2.2.0/gems/eventmachine-1.0.9.1/lib/eventmachine.rb:193:in `run_machine'
/home/deployer/exams/shared/bundle/ruby/2.2.0/gems/eventmachine-1.0.9.1/lib/eventmachine.rb:193:in `run'
/home/deployer/exams/shared/bundle/ruby/2.2.0/gems/faye-websocket-0.10.2/lib/faye/websocket.rb:36:in `block in ensure_reactor_running'
```
/home/deployer/exams/current/cable/config.ru
require ::File.expand_path('../../config/environment',  __FILE__)  
Rails.application.eager_load!
require 'action_cable/process/logging'
ActionCable.server.config.allowed_request_origins = ["https://mydomain"]  
run ActionCable.server

**Comment by rda1902** [comment]
I solved this problem by replacing puma on thin server

**Comment by Tom-Tom** [comment]
Hi, same problem here

@rda1902 : Can you tell me how you have launched ActionCable with Thin ?

**Comment by rda1902** [comment]
@Tom-Tom 
I used this manual https://www.driftingruby.com/episodes/actioncable-on-production

**Comment by carlzulauf** [comment]
Encountering this issue too. Would really like to use Puma + SSL + ActionCable but currently doesn't work. Websocket connections work fine if I turn off SSL.

**Comment by evanphx** [comment]
I’m going to fix this today and have a release out. Thanks for your
patience!

**Comment by evanphx** [comment]
Is there a reason you're running SSL in puma directly for ActionCable?

**Comment by evanphx** [comment]
So I can fix this exact issue but I don't think it will make the system work properly. I'm pretty sure that EM will try to read the encrypted socket but not decrypt it. This is because EM does not use the standard ruby IO mechanism at all (even if Puma used OpenSSL instead of MiniSSL we'd have the same issue, it's not because MiniSSL is in the mix here).

Thusly, I'm going to just have to say that, for now, Eventmachine and Puma + SSL are not compatible.

I'd love to hear more about why people are running Puma + SSL for actioncable rather than running SSL in front of puma though.

**Comment by Tom-Tom** [comment]
@evanphx thanks for this information.

On my side, when i'm on localhost development, i need the SLL everywhere on my rails application.
I can use Thin to run SSL, just for development. But need to rack ActionCable and cannot use the in-app ActionCable like `rails s` with Puma and with route mounting. And Thin is not the example in ActionCable documentation in Rails 5 actually.

It's sure in production i can easily get SSL with NGINX `proxy_pass` and launch Puma or Thin for ActionCable without SSL

The link of @rda1902 is usefull
And if you want NGINX `proxy_pass` :
http://stackoverflow.com/questions/32322652/configuring-nginx-to-proxy-thin-and-rails-actioncable

**Comment by YuriSolovyov** [comment]
Same here. I use puma + ssl + plain rack app running on jruby.
> I'd love to hear more about why people are running Puma + SSL for actioncable rather than running SSL in front of puma though.

It might sound weird, but I do this indeed to drop ssl proxy on front of the app. Nothing to do with `actioncable` though

**Comment by carlzulauf** [comment]
My reason for direct SSL+Puma for action cable was due to an old version of apache that didn't support mod_proxy_wstunnel. I have since switched to a new server with nginx as a workaround. I noticed as well EM seemed to be attempting to manipulate the socket directly and I wasn't able to work around the issue either. I tried making MiniSSL::Socket provide the `fileno` response from it's `@socket` but there are deeper issues it seems.
